### PR TITLE
JS/PY/RB: support a limited number of ranges for ReDoS analysis

### PR DESCRIPTION
**Pull request by erik-krogh**
The idea of this PR is that `/a{0,100}/` is close enough to `/a*/` that we can just treat them the same.   
Gets us one step closer to flagging CVE-2021-22902. 

Evaluations: [Ruby](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/pr-7097-rb__1/reports), [Python](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/pr-7097-py__1/reports), [JavaScript](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/pr-7097-js__1/reports) look OK.  
The JavaScript evaluation shows that an FP was fixed.

### CODE REVIEW

File: ruby/ql/lib/codeql/ruby/regexp/ReDoSUtil.qll [hunk 1]
```diff
@@ -595,6 +595,49 @@ private class EdgeLabel extends TInputSymbol {
...
+/**
+ * A RegExp term that acts like a plus.
+ * Either it's a RegExpPlus, or it is a range {1,X} where X is >= 30.
+ * 30 has been chosen as a threshold because for exponential blowup 2^30 is enough to get a decent DOS attack.
+ */
+private class EffectivelyPlus extends RegExpTerm {
+  EffectivelyPlus() {
+    this instanceof RegExpPlus
+    or
+    exists(RegExpRange range | range.getLowerBound() = 1 and range.getUpperBound() >= 30 |
```
Comment (nickrolfe) [inline ruby/ql/lib/codeql/ruby/regexp/ReDoSUtil.qll:607]: Can you extend this to handle `not exists(range.getUpperBound())` as well? Then we'd catch `{1,}` (1 or more repetitions). Here's [an example in the wild](https://github.com/github/atom.io/blob/66529d5330290cb1fe153e6b5aa6e3b87a3faa3b/app/helpers/sessions_helper.rb#L17).

File: ruby/ql/lib/codeql/ruby/regexp/ReDoSUtil.qll [hunk 2]
```diff
@@ -595,6 +595,49 @@ private class EdgeLabel extends TInputSymbol {
...
+
+/**
+ * A RegExp term that acts like a star.
+ * Either it's a RegExpStar, or it is a range {0,X} where X is >= 30.
+ */
+private class EffectivelyStar extends RegExpTerm {
+  EffectivelyStar() {
+    this instanceof RegExpStar
+    or
+    exists(RegExpRange range | range.getLowerBound() = 0 and range.getUpperBound() >= 30 |
```
Comment (nickrolfe) [inline ruby/ql/lib/codeql/ruby/regexp/ReDoSUtil.qll:621]: Same here.

**Review by yoff — APPROVED** [review]
LGTM

**Review by nickrolfe — APPROVED** [review]

File: ruby/ql/lib/codeql/ruby/security/performance/ReDoSUtil.qll [hunk 3]
```diff
@@ -98,6 +98,8 @@ predicate matchesEpsilon(RegExpTerm t) {
   matchesEpsilon(t.(RegExpBackRef).getGroup())
   or
   forex(RegExpTerm child | child = t.(RegExpSequence).getAChild() | matchesEpsilon(child))
+  or
+  t.(RegExpRange).getLowerBound() = 0
```
Comment (yoff) [inline ruby/ql/lib/codeql/ruby/security/performance/ReDoSUtil.qll:102]: Is this not a repetition of line 86?

Comment (erik-krogh) [reply ruby/ql/lib/codeql/ruby/security/performance/ReDoSUtil.qll:102]: It is.   
I'll remove the redundancy.

**Review by asgerf — APPROVED** [review]
LGTM from JS

**Pull request merged**